pkg/apis/intevents/v1alpha1/brokercell_defaults.go

```diff
@@ -35,22 +63,67 @@ func (bc *BrokerCell) SetDefaults(ctx context.Context) {
 func (bcs *BrokerCellSpec) SetDefaults(ctx context.Context) {
-	if bcs.Components.Fanout.MinReplicas == nil {
-		bcs.Components.Fanout.MinReplicas = ptr.Int32(minReplicas)
+	// Fanout defaults
+	bcs.Components.Fanout.SetCPUDefaults(cpuRequestFanout, cpuLimitFanout)
+	bcs.Components.Fanout.SetMemoryDefaults(memoryLimitToRequestCoefficientFanout, memoryRequestFanout)
+	bcs.Components.Fanout.SetAutoScalingDefaults(targetMemoryUsageCoefficientFanout, avgCPUUtilizationFanout)
+	// Retry defaults
+	bcs.Components.Retry.SetCPUDefaults(cpuRequestRetry, cpuLimitRetry)
+	bcs.Components.Retry.SetMemoryDefaults(memoryLimitToRequestCoefficientRetry, memoryRequestRetry)
+	bcs.Components.Retry.SetAutoScalingDefaults(targetMemoryUsageCoefficientRetry, avgCPUUtilizationRetry)
+	// Ingress defaults
+	bcs.Components.Ingress.SetCPUDefaults(cpuRequestIngress, cpuLimitIngress)
+	bcs.Components.Ingress.SetMemoryDefaults(memoryLimitToRequestCoefficientIngress, memoryRequestIngress)
+	bcs.Components.Ingress.SetAutoScalingDefaults(targetMemoryUsageCoefficientIngress, avgCPUUtilizationIngress)
+}
+
+// SetMemoryDefaults sets the memory consumption related default field values for ComponentParameters.
+func (componentParams *ComponentParameters) SetMemoryDefaults(memoryLimitToRequestCoefficient float64, memoryRequest string) {
+	if componentParams.MemoryRequest == nil {
+		componentParams.MemoryRequest = ptr.String(memoryRequest)
+	}
+	requestedMemoryQuantity := resource.MustParse(*componentParams.MemoryRequest)
+	if componentParams.MemoryLimit == nil {
+		autoSelectedLimit := quantityutil.MultiplyQuantity(requestedMemoryQuantity, memoryLimitToRequestCoefficient)
+		componentParams.MemoryLimit = ptr.String(autoSelectedLimit.String())
+	}
+	// Make sure the limit is not lower than what's requested
+	memoryLimitQuantity := resource.MustParse(*componentParams.MemoryLimit)
+	if memoryLimitQuantity.Cmp(requestedMemoryQuantity) < 0 {
+		componentParams.MemoryLimit = ptr.String(requestedMemoryQuantity.String())
+		memoryLimitQuantity = requestedMemoryQuantity
+	}
+}
+
+// SetCPUDefaults sets the CPU consumption related default field values for ComponentParameters.
+func (componentParams *ComponentParameters) SetCPUDefaults(defaultCPURequest, defaultCPULimit string) {
+	if componentParams.CPURequest == nil {
+		componentParams.CPURequest = ptr.String(defaultCPURequest)
+	}
+	if componentParams.CPULimit == nil {
+		componentParams.CPULimit = ptr.String(defaultCPULimit)
 	}
-	if bcs.Components.Fanout.MaxReplicas == nil {
-		bcs.Components.Fanout.MaxReplicas = ptr.Int32(maxReplicas)
+}
+
+// SetAutoScalingDefaults sets the autoscaling-related default field values for ComponentParameters.
+func (componentParams *ComponentParameters) SetAutoScalingDefaults(targetMemoryUsageCoefficient float64, avgCPUUtilization int32) {
+	if componentParams.MemoryLimit == nil {
+		panic("Memory limit should be specified on the component before setting auto-scaling parameters")
 	}
```

Comment:
Try avoid `panic` as much as possible. In this case we definitely don't want to crash the webhook for a single invalid value. Instead, valid it in `*_validation.go`.

Reply:
Also notice `MustParse` will also cause panic.

Reply:
The order of webhook is defaulting first then validation. So if you see an invalid value here, simply pass it along and let the validation webhook reject the request.

Reply:
A bit more context here: this was used in a 'code contract' way, i.e. the methods were structured such that this condition is not triggerable externally and was rather meant to set input expectations internally. I see that there are more suitable practices here and the latest updates should be more aligned with those.